### PR TITLE
Improving evalasdata

### LINKED ISSUE #116: [Lua datamodel] Wrong behaviour if function assigned to <data> element
Hello, after assigning function to \<data\> element its content is evaluating. I think that this is wrong behaviour

```
<scxml datamodel="lua" initial="StateShape1" name="ScxmlShape1" version="1.0" xmlns="http://www.w3.org/2005/07/scxml">
	<datamodel>
		<data expr="0" id="Var1"/>
		<data id="ChangeVar1">function ()
    print(&quot;We must not see it!&quot;)
    Var1 = 555
end
		</data>
	</datamodel>
	<state id="StateShape1">
		<transition cond="Var1==0" target="Pass"/>
		<transition cond="Var1==555" target="Fail"/>
	</state>
	<final id="Pass"/>
	<final id="Fail"/>
</scxml>
```

### LINKED ISSUE #118: [Lua datamodel] Function 'evalAsData' requires correction
Hello, 
1)I am not fully satisfied that content executes a couple of times. You are trying to do everything by 1 function. Why not to create separate functions such as evalExpr, evalScript?

2)Imagine, that we have very difficult script and trying to debug it. 
```
<scxml datamodel="lua" initial="Main" name="ScxmlShape1" version="1.0" xmlns="http://www.w3.org/2005/07/scxml">
	<state id="Main">
		<onentry>
			<script>local s_text=&quot;xxx&quot;


-- a lot of code


validVar = &quot;zzz&quot;
nonValidVar
			</script>
		</onentry>
		<transition event="error.*" target="FinalShape1"/>
	</state>
	<final id="FinalShape1"/>
</scxml>
```
After executing its content we see next message:
`"cause": "[string \"return(local s_text=\"xxx\"...\"]:1: unexpected symbol near 'local'",`
And we can not understand, what's happend in real.

But if we called it without 'return' we would see next:
`"[string \"local s_text=\"xxx\"...\"]:8: '=' expected near '<eof>'"`
In this case we can very easy find where the problem is

### LINKED ISSUE #119: Suggestion to make SCXMLIOProcessor also exported in DLL
Hello, it would be great if you made SCXMLIOProcessor also exported.

`class USCXML_API SCXMLIOProcessor : public IOProcessorImpl`

### LINKED ISSUE #120: Logger in ActionLanguage is not copied to invoking sessions
Hello, look at USCXMLInvoker.cpp

```
void USCXMLInvoker::invoke(const std::string& source, const Event& invokeEvent) {
...
ActionLanguage* alOrig = _callbacks->getActionLanguage();
		if (alOrig != NULL) {
			ActionLanguage al;
			// create new instances
			al.execContent = alOrig->execContent.getImpl()->create(invoked);
			al.delayQueue = alOrig->delayQueue.getImplDelayed()->create(invoked);
			al.internalQueue = alOrig->internalQueue.getImplBase()->create();
			al.externalQueue = alOrig->externalQueue.getImplBase()->create();
			al.microStepper = alOrig->microStepper.getImpl()->create(invoked);
			_invokedInterpreter.setActionLanguage(al);
		}
```

I think that next code is missed
`al.logger = alOrig->logger.getImpl()->create();`

### LINKED ISSUE #124: [Lua datamodel] Test 179 is not passed
Hello, we can not pass test 179 with LuaDataModel, because in this test we are sending event with content that is number value equal to 123, then we converting it to string because of the code in function **BasicContentExecutor::elementAsData**
```
if (LOCALNAME(element) == "content") {
	return Data(spaceNormalize(contentSS.str()), Data::VERBATIM);
```
And of course we can't pass the condition
`_event.data==123` because _event.data is string and equal "123", so `"123"~=123`
that's why first of all we need to check wether content is a valid data value
```
// need first try getAsData because how to pass 179 ?
 try {
	// test153, we need to throw for test150 in promela
	Data d = _callbacks->getAsData(contentSS.str());
	if (!d.empty())
	        return d;
}
catch (ErrorEvent &) {
	// test294, test562       
       return Data(spaceNormalize(contentSS.str()), Data::VERBATIM);
}
```

### PR DESCRIPTION
**Pull request by alexzhornyak**
1) assign Logger via setActionLanguage (fixes #120)
2) export SCXMLIOProcessor in DLL (fixes #119)
3) implemented function evalAsScript (fixes #118)
4) corrected BasicContentExecutor to pass tests 179, 294, 562 (fixes #124)
5) fixes #116

### CODE REVIEW

**Comment by sradomski** [comment]
Could you be so kind as to split this push request into four individual ones. I do not have any problem with `1. assign Logger via setActionLanguage`, which is obviously correct. Neither do I object to `2. export SCXMLIOProcessor in DLL`. But your `3. implemented function evalAsScript` seems lacking and I do not see why you would introduce it in the first place. The idea is that the various `evalAsX` take an expression, a statement or a sequence thereof of the target language, evaluate them in the respective datamodel and return any eventual result. Here, `evalAsBool` is an optimization for `cond` attributes of transition elements in order not to construct a complete `Data` object. As such, your `evalAsScript` is in essence yet another optimization were no return value is expected at all, but you do not provide any implementations for the other datamodels. Furthermore, I was not aware that `4. corrected BasicContentExecutor to pass tests 179, 294, 562` is necessary, as those tests pass before your introduction of `evalAsScript`.

TL;DR: I have no issue with 1. and 2. but object to the introduction of `evalAsScript` as its functionality is already available in `evalAsBool` and optmization of not even returning a `bool` seem minuscule when compared to the additional maintenance effort for every datamodel.

**Comment by alexzhornyak** [comment]
1) about evalAsScript
I will try to explain. I think that in LuaDataModel this is wrong to evaluate script content like returning expresion how that is implemented in evalAsData, evalAsBool as
`int retVals = luaEval(_luaState, "return(" + trimmedExpr + ")");`
I explained that this form of evaluating gives wrong behaviour when you have syntax errors in script, etc.
Thus debugging of difficult scripts is almost impossible. Also we loose performance because we need to execute code twice. So, evalAsData in LuaDataModel seems to be function with hacks that is not clear for understanding.
So, I am sure that for Lua datamodel we must understand when we are evaluating expression that can return something and when we are evaluating script.
` luaEval(_luaState, trimmedExpr );`
If you think that evalAsScript function is redundant and not necessary for other datamodels then maybe pass optionally the param to evalAsData like
```
typedef enum {edtExpr, edtScript} EvalDataType;
virtual Data evalAsData(const string &sExpr, const EvalDataType AType = edtExpr);
```
and than datamodel will decide whether use it or not. 
So, for Lua I can not leave as is, and if we are talking of development of really difficult project we must do separate it.
2) about tests 179, 294, 562
Without my corrections of BasicContentExecutor you can not pass test 179 with LuaDataModel, because in this test we are sending in event content that is number value equal to 123, then in old code we converting it to string
```
if (LOCALNAME(element) == "content") {
	return Data(spaceNormalize(contentSS.str()), Data::VERBATIM);
```
And of course we can't pass the condition
`_event.data==123` because _event.data is string and equal "123", so `"123"~=123`
that's why first of all we need to check wether content is a valid data value

3) and what about function issue 116, you didn't tell about your opinion

**Comment by sradomski** [comment]
If you agree, I would close this as most of the changes are already in the master branch and the evalAsData discussion can continue in issue 118.
